docs: create the model generation guide

Florence-Njeri opened the pull request:
<!--   Thank you for your contribution. Before you submit the pull request:
1. Follow contributing guidelines and make sure PR title follows Conventional Commits specification -> https://github.com/asyncapi/generator/blob/master/CONTRIBUTING.md
2. Test your changes and attach their results to the pull request.
3. Update the relevant documentation.
-->

**Description**
- New guide explaining how to go about generating types using Modelina
- In the introduction and the template development docs, I've added a note stating that if someone is only interested in types/model generation, they should use Modelina directly.
- Updated CLI naming

**Related issue(s)**
Resolves https://github.com/asyncapi/generator/issues/1042

derberg commented on docs/model-generation.md:
```diff
@@ -0,0 +1,128 @@
+weight: 200
+---
+
+Suppose you want to generate models for your application from the data types you have using the asyncapi generator. Then you should use Modelina instead of the asyncapi generator with a template.
```
```suggestion
Use [AsyncAPI Modelina](/tools/modelina) to generate models for an application. Since models generation is something that you usually perform as an automated step through CI/CD, best way to use Modelina is through [AsyncAPI CLI](/tools/cli).

In case you work on a template for generating an application and part of template responsibilities is to provide models, you should integrate Modelina library instead of templating model structures using features from AsyncAPI Generator.
```

derberg reviewed (changes requested):
You will need to adjust narrative in the main document as the scope of https://github.com/asyncapi/generator/issues/1042 was not a tutorial, but more of a guide that tells you what is needed and shows an example. 

- we do not need to mention https://github.com/derberg/python-mqtt-client-template. You can just add a note that the example you provide can be integrated in a template that one can create by following https://www.asyncapi.com/docs/tools/generator/generator-template
- I think it is enough that first step of the guide just says that `@asyncapi/modelina` needs to be added to template dependencies as `npm install --save @asyncapi/modelina` because then it is automatically added to `package.json`
- `Model generation using the AsyncAPI CLI` section can be removed as it is not adding anything new, template with or without modelina - generation is always the same. You can just link to https://www.asyncapi.com/docs/tools/generator/usage#generator-cli (btw we have error there as section should be called AsyncAPI CLI, and we should also have a link to https://www.asyncapi.com/docs/tools/cli there)
- please remember about proper linking inside asyncapi.com with relative links and also consistent usage of upper/lower cases with names of tools




also, in your PR can you add between line 37 and 38 in https://github.com/asyncapi/generator/blob/master/.github/workflows/update-docs-in-website.yml#L37-L38 a line `rm -r ./pages/docs/tools/generator` to always have a clean and empty state of generator docs folder in website? Problem is that we now do not remove, and just add, so when we rename files like this https://github.com/asyncapi/generator/blob/master/docs/generator-template.md then we now get a duplicated tutorial in website
![Screenshot 2024-01-04 at 14 24 49](https://github.com/asyncapi/generator/assets/6995927/0248a4eb-038d-4373-a42e-bc1bb0913466)

Florence-Njeri commented:
> You will need to adjust narrative in the main document as the scope of #1042 was not a tutorial, but more of a guide that tells you what is needed and shows an example.
> 
> * we do not need to mention https://github.com/derberg/python-mqtt-client-template. You can just add a note that the example you provide can be integrated in a template that one can create by following https://www.asyncapi.com/docs/tools/generator/generator-template
> * I think it is enough that first step of the guide just says that `@asyncapi/modelina` needs to be added to template dependencies as `npm install --save @asyncapi/modelina` because then it is automatically added to `package.json`
> * `Model generation using the AsyncAPI CLI` section can be removed as it is not adding anything new, template with or without modelina - generation is always the same. You can just link to https://www.asyncapi.com/docs/tools/generator/usage#generator-cli (btw we have error there as section should be called AsyncAPI CLI, and we should also have a link to https://www.asyncapi.com/docs/tools/cli there)
> * please remember about proper linking inside asyncapi.com with relative links and also consistent usage of upper/lower cases with names of tools
> 
> also, in your PR can you add between line 37 and 38 in https://github.com/asyncapi/generator/blob/master/.github/workflows/update-docs-in-website.yml#L37-L38 a line `rm -r ./pages/docs/tools/generator` to always have a clean and empty state of generator docs folder in website? Problem is that we now do not remove, and just add, so when we rename files like this https://github.com/asyncapi/generator/blob/master/docs/generator-template.md then we now get a duplicated tutorial in website ![Screenshot 2024-01-04 at 14 24 49](https://private-user-images.githubusercontent.com/6995927/294191595-0248a4eb-038d-4373-a42e-bc1bb0913466.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDQ4MjIzMzEsIm5iZiI6MTcwNDgyMjAzMSwicGF0aCI6Ii82OTk1OTI3LzI5NDE5MTU5NS0wMjQ4YTRlYi0wMzhkLTQzNzMtYTQyZS1iYzFiYjA5MTM0NjYucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDEwOSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDAxMDlUMTc0MDMxWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NGY4MDk0MzQ0N2UxYWU2OTk3MzBkZGY3MDg3YzJkYTM5NzAzNWY2NTA0MGQyNTk0NjM3NWY0NzAzYmYyZjJiZSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.d_BtsCU25SijWFwd_u3MOoHPA9KnzYA90CQGuQKRwXY)

Hey @derberg, I have addressed all the feedback you left on this PR

Florence-Njeri commented on docs/usage.md:
```diff
@@ -43,7 +43,7 @@ npm install <folder>
 ### Global templates installed with `yarn` or `npm`
 
-You can preinstall templates globally before installing the generator CLI. The generator first tries to locate the template in local dependencies; if absent it checks where the global generator packages are installed.
+You can preinstall templates globally before installing the [AsyncAPI CLI](https://www.asyncapi.com/docs/tools/cli). The generator first tries to locate the template in local dependencies; if absent it checks where the global generator packages are installed.
```
@derberg I'm not sure if this is the right place to link to the CLI tool docs. Let me know, what you think

derberg replied:
lgtm

Florence-Njeri commented:
@derberg this is ready for a second review

Florence-Njeri commented on docs/model-generation.md:
````diff
@@ -0,0 +1,108 @@
+  // 4 
+  const pythonGenerator = new PythonGenerator();
+// 5
+  const models = await pythonGenerator.generate(asyncapi);
+// 6
+  const files = [];
+
+  for (const model of models) {
+    // 7
+    const modelFileName = `${FormatHelpers.toPascalCase(model.modelName)}.py`;
+    // 8
+    files.push(<File name={modelFileName}>{model.result}</File>);
+  }
+// 9
+  return files;
+}
+```
+
+Let's break it down. The code snippet above does the following:
+
+1. The **File** component from the [generator react SDK](https://github.com/asyncapi/generator-react-sdk) is imported. This tells the AsyncAPI CLI to use the [react render engine](https://www.asyncapi.com/docs/tools/generator/react-render-engine) for generating model files.
+2. Modelina is imported into your template. The Modelina library makes the **PythonGenerator** and the **FormatHelpers** for formatting the generated Python code.
+3. An asynchronous function **schemaRender**, is defined and is invoked everytime the AsyncAPI CLI runs the **models.js** template. This function is tasked with rendering the generated schema files and takes the `asyncapi` object as an input parameter.
+4. A new instance of the **PythonGenerator** model generator from Modelina is created and is used to generate Python code from the provided AsyncAPI document.
+5. An array of Python model classes is generated from the `asyncapi`,AsyncAPI document, object received from the generator.
+6. An empty array **files** is defined to store the generated model files.
+7. The for loop iterates over each generated model and formats the generated model names to Pascal case.
+8. The filenames and the model schemas are added to the **files** array.
+9. The function returns an array of files, each representing the generated model class, to the generator engine.
+
+>Note:
+>You can integrate this example into your own template by following the [generator template tutorial](https://www.asyncapi.com/docs/tools/generator/generator-template). Ensure that you add `@asyncapi/modelina` to template dependencies using `npm install --save @asyncapi/modelina`.
+
+## Generate models
+
+To learn how to generate models using the AsyncAPI CLI, refer to the section on Generator CLI usage. Execute the command provided there, and upon successful generation, the output will guide you to the location of your newly created model files in the specified directory **test/project/src/models**.
+
+## Conclusion
+
+Modelina provides a flexible and powerful way to generate data models from AsyncAPI, OpenAPI, or JSON Schema documents. With the AsyncAPI CLI, you can easily generate models and integrate them into your development workflow. With the ability to customize the generated models, Modelina proves to be a valuable tool in the development of event-driven architectures.
````
Taking a look

Florence-Njeri reviewed (commented):
Review from Lukasz

derberg commented:
@Florence-Njeri please have a look at this comment with big modification - there was suppose to be much more content, not just finish with adding dependency 🤔 

also do not forget about 👇🏼 
> please also update the title and description of PR - we do not merge tutorial with this PR

derberg commented:
sorry @Florence-Njeri my comment was weirdly interpreted by GitHub commit suggestion

pasting the same comment below:




---
title: "Adding models generation in template"
weight: 200
---

This guide will walk you through the process of enabling models/types generation in a template by using [Modelina](https://www.asyncapi.com/tools/modelina).

Modelina is an AsyncAPI library designed for generating data models by utilizing inputs like [AsyncAPI](generator/asyncapi-document), OpenAPI, or JSON schema inputs. Its functionality revolves around creating data models from the provided AsyncAPI document and the model template, which defines message payloads. It is better to use Modelina in your template to handle model generation rather than providing custom templates.

You can integrate work shown in this guide in a template that you can create by following [tutorial about creating a template](https://www.asyncapi.com/docs/tools/generator/generator-template).

In this guide, you'll learn how to use Modelina in a template code to enable support for Python data models generation. 

## Add Modelina dependency

Install Modelina in your project using npm: `npm install --save @asyncapi/modelina`.

Your template's `package.json` file should now contain Modelina pointing to its latest version:

  ```json
  "dependencies": {
     ...
     "@asyncapi/modelina": "^2.0.5"
     ...
   }
  ```

## Create a models.js file

Create a new directory in the **template** directory named **src/models** and create a **models.js** file within it. In the **models.js** file, add the following code:

```js
// 1
import { File } from '@asyncapi/generator-react-sdk';
// 2
import { PythonGenerator, FormatHelpers } from '@asyncapi/modelina';

/**
 * @typedef RenderArgument
 * @type {object}
 * @property {AsyncAPIDocument} asyncapi document object received from the generator.
 */

/**
 * Render all schema models
 * @param {RenderArgument} param0 
 * @returns 
 */
// 3
export default async function schemaRender({ asyncapi }) {
  // 4 
  const pythonGenerator = new PythonGenerator();
  // 5
  const models = await pythonGenerator.generate(asyncapi);
  // 6
  const files = [];
  // 7
  for (const model of models) {
    // 8
    const modelFileName = `${FormatHelpers.toPascalCase(model.modelName)}.py`;
    // 9
    files.push(<File name={modelFileName}>{model.result}</File>);
  }
  return files;
}
```

Let's break it down. The code snippet above does the following:

1. The **File** component from the [generator react SDK](https://github.com/asyncapi/generator-react-sdk) is needed to handle further rendering of generated models into files.
2. The **PythonGenerator** generator is a core that you need for models generation. Additionally you can import [**FormatHelpers**](https://github.com/asyncapi/modelina/blob/master/src/helpers/FormatHelpers.ts) that give you a set of helpers that make it easier to modify model names to match your required case.
3. You can change the name **schemaRender** to anything else like **modelRenderer**. More important is that this must be an `async` function and be a default export. This function is invoked during generation process and it should contain the logic behind models generation.
4. First you create an instance of the **PythonGenerator** model generator. If you decide to use present functionality from Modelina, you need to pass your presets here during instance creation.
5. Actual models generation is one line of code, and as a result you get an array of models that later you need to turn into files.
6. You need to define an array that must be returned from **schemaRender** function. The array must contain React components, and in this case `<File>` component.
7. Iterate over generated models and use their content to create proper definitions of `<File>` component.
8. Notice how using Modelina helpers, in this case **toPascalCase** function, you can make sure that the filename of your model follows specific case pattern.
9. Each component must be added to **files** array that you later return from the default function. Notice the definition of `<File>` component that enables you to provide the name of resulting file and the content of the model. Notice also `model.result` that shows that initially generated array with models did not contain raw models content but a set of output objects that contain not only `result` but also other info, like for example `modelName`.

With such model template that uses Modelina, as a result of generation process you would receive a set of model files in `$OUTPUT_DIR/src/models` directory.

## Conclusion

Modelina provides a flexible and powerful way to generate data models from AsyncAPI, OpenAPI, or JSON Schema documents. By integrating Modelina you can much faster enable models generation in your template. 
```

suggested restructure and rewrite:
- give more details
- follow guide not tutorial
- remove CLI as in this topic CLI is not relevant, especially that Generator can also be used as library and not only as tool integrated in AsyncAPI CLI

@Florence-Njeri I suck at grammar, so please after accepting the change, do a review and probably improvement
```

derberg reviewed (approved):
@jonaslagoni @kennethaasan @magicmatatjahu since this is Modelina related, anything from your side? is it a good way to promote Modelina as component that enables models generation in app template?

maybe long term would be better to have a dedicated React component that makes it even easier to embed model generation - but for now I think it should be enough, especially that there is nothing else at the moment

jonaslagoni reviewed (approved):
LGTM 👍

derberg commented:
/rtm

aeworxet commented:
/bounty
